### ISSUE TITLE
History and blame window do not show merge commits

### ISSUE THREAD

**Issue by aholeton** [ForkIssues/TrackerWin#1261]
When a branch is merged into another branch, the merge commit is not reflected in the history or blame window for the file. I would expect the blame and history window to show all commits that affected this file, but merge commits are omitted from the view. Some merges are complicated, so having these omitted from the history makes the history incomplete and hard to reason about what happened to the file. I'm not sure if this is intentional or a bug. However, there should be an option of showing these types of commits.

Example - The commit "Merge Branch1" is omitted from the history.

![Example](https://user-images.githubusercontent.com/64423411/127937033-f47dd4ff-b493-4077-b54c-84ab5f7cb53c.png)

**Comment by ShadowZzj** [comment]
Have the problem solved? I have encounter the same issue, when I use git command "git log --pretty-oneline <filename>"，the log shows the merge commit that change the file. However, In fork, it doesn't.

**Comment by anrichter** [comment]
Unfortunateley the problem still exists in version 1.74.1.0. 
Is the fix planned for one of the next version?

**Comment by ofthelit** [comment]
I have also encountered this bug. I noticed that the latest version of the file in the File Tree has some changes that are not visible in the latest change of the History view. This is because those changes were included in the merge commit.

The history view of Visual Studio shows the merge commit.

I'm on Fork 1.86.1.0.

It won't happen too often when one encounters this bug, but it is still an important issue to fix since it hides info.

**Comment by duoslav** [comment]
Still the same behavior v1.89.2.0.

Not really crucial but would be nice to fix

**Comment by Cyame** [comment]
got an idea/hint for that, see if it solve the problem.

I got a repository which have change in this file in the merge node, 

I use this command
```
git log --follow -c -- <file-name>
```
it cannot track the merge node's log.

but using
```
git log --follow --full-history --simplify-merge -c -p -- <filename>
```
could show this merge editing node.

See if it helps enhance this feature.

**Comment by rutgersc** [comment]
Agree, was missing crucial changes by using the history. Curious what other tools you guys use in the meantime to view the full history ? Sourcetree maybe?

**Comment by JonUleis** [comment]
This has tripped me up a few times and a fix would be helpful. Fork's History view for a file is sometimes confusingly incomplete if changes occurred during a merge

**Comment by Srekel** [comment]
I spent about an hour yesterday being confused trying to track down where a certain change to a file happened and eventually realized it was in a bad merge conflict commit. Had the commit been in the history and the blame window it would probably have found the problem instantly. And this isn't the first time this has happened, so please prioritize this :)

**Comment by bitzer-mh** [comment]
Fork, please come on, please :-)
A files content can be changed without it being visible in the History window.
Is that not a bug that needs to be resolved?
At least an option to select Full/Simplified etc. history display, like e.g. on the Azure web...

**Comment by mucco** [comment]
I also sorely missed this feature today, made my work a lot harder finding a culprit merge commit. Please fix it!

**Comment by Hykke** [comment]
I just had an issue where there were no changes to the file after the merge, so now I could not get a blame for how the file currently looks, I could only get a blame of how the file looked before the merge, either on the feature branch or from the previous commit on the main branch - This was not obvious to me, and I spent an hour trying to figure out why the file I was seeing in blame was not equal to the file I could see in my text editor.

**Comment by goran-w** [comment]
The many different "Commit Limiting" and "History Simplification" option flags for `git-log`, applied to specified file/directory path(s), are interacting in some intricate ways...

The "Default mode" for History Simplification will ONLY show the MINIMAL set of commits needed to explain the CURRENT state (i.e at the specified commit(s), or HEAD). So, for example, if a certain change was introduced in one commit and then undone in a later commit, both of these commits may be skipped from the resulting log (unless there were other changes in the same file). Also, the log may include the commit where a certain change was originally made, but not a later merge where that change was "trivially" introduced into some branch.

I did some experimenting with the available options and found the following:
* The `--oneline` (formatting) option helps make the log listing compact when experimenting.
* The `--graph` (formatting) option is useful for making more sense of non-linear history, but be aware that this option also has the side-effect of enabling "parent rewriting" which may alter the effects of other options below!
* The `--follow` (general) option is supposed to "continue listing the history of a single file beyond renames", but it has other subtle side-effects and does not work well in combination with other options. Most importantly, as briefly hinted on the manual page for the `log.follow` git-config, it "does not work well on non-linear history" - it may alter the effects of other options discussed here, so use it with caution...
* The `--simplify-merges` (history simplification) option can be used standalone or in combination with some of the other options. It is useful for getting a more detailed (non-linear) history, but may list merge-commits that don't actually change the specified file(s).
* The `--full-history` (history simplification) option usually gives too many commits, so I wouldn't bother with it (unless I really need to see ALL commits where the file was changed on ANY branch). NOTE: Avoid using this option with "parent rewriting" (via `--graph`, `--parents` or `--children`), since it may produce a huge log (unless we're also specifying `--simplify-merges`).
* The `--show-pulls` (history simplification) option is useful to show merge commits that "first introduced" a change to a branch. However, it needs to be combined with the more "noisy" option `--simplify-merges` to show every related merge (possibly including some where the file was not actually changed). NOTE: The latter combo surprisingly shows more commits than when using `--full-history --simplify-merges`!
* The `--first-parent` (commit limiting) option is very useful for understanding how a specific file has evolved on a specific branch, while keeping the history log more compact! With this option (used on its own) we can see all the commits (including merges) along the first-parent ancestry-line which actually changed the specified file. (NOTE: We don't, however, see the commits where these merged changes were originally introduced on another branch.)

To summarize the above into a couple of recommendations (add more output formatting flags as needed) :
* For a rather complete non-linear history log, including the most relevant merges (even some with no changes), use:
  * `git log --show-pulls --simplify-merges -- <file>`.
* For a very compact linear history log, which does not skip any changes along the first-parent ancestry-line, use:
  * `git log --first-parent -- <file>`.

On a final note, it might be interesting (but possibly dificult) to extend the History view with a log-graph (similar to the one produced by the `--graph` option, for non-linear history).

**Comment by goran-w** [comment]
BTW, another way to include merge-commits in a file-history `git-log` is to specify the `-m` or `--diff-merges=on` (diff formatting) options. (These seem to produce the same, or at least similar, results as when specifying the `--full-history` option.)
 
In Azure DevOps Services, there's a drop-down of related options for selecting the level of details in their History view:
<img width="1876" height="420" alt="Image" src="https://github.com/user-attachments/assets/7471b1ad-9f95-44b2-b77e-c91c344b9e05" />
https://devblogs.microsoft.com/devops/announcing-git-graph-and-advanced-filters-to-visualize-commit-history/#advanced-filters

**Comment by bilbothebaggins** [comment]
I hit the same problem in Fork while validation some merge use cases.
As far as I can see, the problem is that Fork hardcodes the switches when invoking `History` or `Blame/Timeline` on a file.

I have 2.15.4.0 at the moment, and Process Monitor shows this command line:
```
git.exe -c core.quotePath=false log 
  --no-show-signature 
  --oneline 
  --name-status 
  --follow 
  --pretty=format:%H%n%aN%n%aE%n%ct%n%s 
  0a24558a74678d027bc9bf13fd58a3c89620f17a 
  -- "E.txt"
```
when I try to show the history on my test file.

All other tools I have tried today (Sourctree, sourcegit, GitLens) work, because they do not hardcode the switches (specifically `--name-status` `--follow` I think).

That is, while **git itself** is partially to blame here, namely https://stackoverflow.com/a/35786883 from [#1580](https://github.com/fork-dev/Tracker/issues/1580), **Fork** seems to be unique in that it passes both name-status and follow unconditionally to it's history commands.

I think Fork should reconsider that choice or add an option to control this.

References:
* https://github.com/fork-dev/Tracker/issues/1580
* https://github.com/sourcegit-scm/sourcegit/issues/2044
* https://stackoverflow.com/questions/35780773/how-to-get-git-log-name-status-to-work-with-merge-commits/35786883#35786883

**Comment by bilbothebaggins** [comment]
I will add another observation here:

While OP shows an example where history/blame omits intermediate commits (bad enough IMO) -- Fork will ALSO omit the head commit, if that is a merge commit.

That is, given
```
* Latest <<
|  \ 
|   * Second Parent
|   | 
|   ...
|
* First Parent
|
...
```

... if I do the following in Fork 2.22 on Windows:
* Select `Latest` in graph
* Switch to _Changes_ Tab on the details
* Open _Blame/Timeline..._ (or _History_) for a file of this merge commit

... then the Blame window of Fork will **not even show the content of the commit/file I selected** in the GUI. It will show the content of the last history entry it has (the last non merge commit in first parent order).

I consider skipping over older history entries bad enough.
Opening a Blame window on a specific commit-file from the UI and then _not_ showing the content of that commit in the Blame window, but the content of a different, older, commit, is totally breaking common user expectations for me.

Fork is an excellent tool, but the Blame window is broken IMO.

Maybe there's hope for an option where the underlying git history command can be tweaked by the user, even if you think the current default is useful? :-)

**Comment by Srekel** [comment]
There is a workaround that I have started to use. You can get a full list of commits in the search window (though this probably doesn't work for file renames..?)

It shows both the commits and the merges when they went in, which is quite useful.

<img width="529" height="305" alt="Image" src="https://github.com/user-attachments/assets/66966353-c710-455a-a239-4974e3b36797" />

<img width="434" height="283" alt="Image" src="https://github.com/user-attachments/assets/dee844fa-dfa6-42d7-abb6-4f4807997155" />
